### ISSUE TITLE
Trainer Stuck at 0% Progress during Training on Multi-GPU Setup

### ISSUE THREAD

**Issue by yanho824** [huggingface/transformers#38008]
### System Info

- transformers version: 4.37.2
- Platform: Linux version 5.15.0-105-generic
- Python version: 3.10.17
- Huggingface_hub version: 0.31.1
- Safetensors version: 0.5.3
- Accelerate version: 0.26.0
- PyTorch version (GPU?): 2.1.2+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Machine: 4 x A800 GPUs

### Who can help?

@ArthurZucker 
@zach-huggingface
@SunMarc

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I am running the following script:

> CUDA_VISIBLE_DEVICES=0,1,2,3` python svd_trainer.py --model_id meta-llama/Llama-2-7b-hf --target_ratio 0.4 --remapping

The core parts of the code that may be related to the issue are:

>  
   # training
    data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)
    from transformers import TrainingArguments
    training_args = TrainingArguments(
        output_dir= TA_tarined_model_output_dir,
        num_train_epochs = TA_num_train_epochs,
        evaluation_strategy = "epoch",
        per_device_train_batch_size = 1,
        per_device_eval_batch_size = 1,
        warmup_steps = TA_warmup_steps,
        lr_scheduler_type = "cosine",
        seed = args.seed,
        gradient_accumulation_steps = TA_gradient_accumulation_steps,
        # load_best_model_at_end = True,
        save_strategy = "no",
        save_steps = 1000,
        save_total_limit = 2,
        remove_unused_columns=False,
        # deepspeed =deepspeed_config
    )
   
    trainer = SVDTrainer(
        model=model,
        args=training_args,
        train_dataset=tokenized_traindata,
        eval_dataset=tokenized_valdata,
        data_collator=data_collator,
    )

    
    model,  train_dataloader, eval_dataloader = accelerator.prepare(
        model, trainer.get_train_dataloader(), trainer.get_eval_dataloader()
    )
    trainer.train = accelerator.prepare(trainer.train)
    
    
    # save setting
    TA_tarined_model_output_dir.mkdir(parents=True, exist_ok=True)
    output_json_path = TA_tarined_model_output_dir / "para_config.json"
    
    data = {
        "model_id": model_id,
        "DATASET_NAME": DATASET_NAME,
        "target_compression_ratio": target_compression_ratio,
        "BETA": BETA,
        "dataset_processing": {
            "SEED": args.seed,
            "SEQ_LEN": SEQ_LEN,
            "NSAMPLES_train": NSAMPLES_train,
            "NSAMPLES_val": NSAMPLES_val,
        },
        "training_settings": {
            "TA_num_train_epochs": TA_num_train_epochs,
            "TA_warmup_steps": TA_warmup_steps,
            "TA_gradient_accumulation_steps": TA_gradient_accumulation_steps
        },
        "loss_function_settings": {
            "lambda_reg": lambda_reg
        },
        "optimizer_settings": {
            "scheduler_lr": scheduler_lr,
            "scheduler_step_size": scheduler_step_size,
            "scheduler_gamma": scheduler_gamma
            # "scheduler_half_step_size": scheduler_half_step_size,
            # "scheduler_divide": scheduler_divide
        },
        "dtype_settings": {
            "model_load_dtype": "f16",
            "computeSVD_dtype": "f32"
        },
        "no_svd_layer": model_no_svd_layer_dic[lower_id],
    }
    with open(output_json_path, "w") as json_file:
        json.dump(data, json_file, indent=4)

    
    print("training start, target_compression_ratio is ",target_compression_ratio, " , NGPUS is ", NGPUS)
    trainer.train()
    
    accelerator.wait_for_everyone()

It stuck at 0% more than 12 hours.

![Image](https://github.com/user-attachments/assets/f674c46e-a55b-4bc1-84cc-446845efcce4)

I wrote a test code to determine that Trainer can run on a single GPU but not on multiple GPUs. However, the program will report an error after disabling P2P by referring to the solution in https://github.com/huggingface/transformers/issues/26724.

![Image](https://github.com/user-attachments/assets/92b773e5-81c0-4a3a-8f6d-52a0782e4ca6)

Here's how I checked the communication between the GPUs after disabling P2P

![Image](https://github.com/user-attachments/assets/c1d7271c-1ec8-49ff-bdbd-1dcd8b1c36e8)

### Expected behavior

If you have any insight or suggestions on how to resolve this issue, we would appreciate it.Thanks!

**Comment by htdung167work0** [comment]
I got this error too. It appears when load_best_model_at_end = True on multi-gpus training.

**Comment by SunMarc** [comment]
Hey @yanho824, where does this SDVtrainer comes from ? Is you subclassed Trainer, you don't need the following as we take care of that in Trainer. Also please share the full minimal reproducer  : 
```python
model,  train_dataloader, eval_dataloader = accelerator.prepare(
    model, trainer.get_train_dataloader(), trainer.get_eval_dataloader()
)
trainer.train = accelerator.prepare(trainer.train)
```
